### PR TITLE
Implement the BGG resolution of a simple finite dimensional Lie algebra representation

### PR DESCRIPTION
**Pull request by tscrim**
The goal of this PR is to implement the BGG resolution of a finite dimensional simple Lie algebra representation $L_{\lambda}$. In order to implement this, we need to implement a number of features:

- Simple modules (this does not assume finite dimensional).
- Dual modules in BGG category $\mathcal{O}$.
- Fix embeddings of Verma modules #36793.
- Improve construction of elements of free abelian monoids when strings are indices.
- Move `to_root_vector()` to a method of fundamental weights and cache the inverse Cartan matrix.
- Add `is_dominant_weight()` method to weight lattice realization elements.
- Add `is_verma_dominant()` to define a dominant weight in the sense of the Humphreys 2008 reference, which means the Verma module is projective (that is, maximal in its dot action orbit).
- Add the category of `FiniteDimensionalKacMoodyAlgebras`.
- `@cache_in_parent_method` marked `weak_le` for Coxeter groups.
- Subclass of PBW bases for semisimple Lie algebras.

We also add methods to compute the contravariant form of a Verma module and add tester methods for them being simple or projective.

It is possible to split this up into smaller parts, but it is natural for them to go together given the final goal.

One major TODO is to speed up the computation of the bases of the simple modules. The main place to optimize this would be to improve the multiplication of the PBW basis code, which for finite dimensional Lie algebras (not necessarily semisimple), this could be improved by using the fixed basis order and using the monomials as vectors to avoid lots of calls for getting the sort order and comparing. In particular, the triangular sorting order for the Chevalley basis implementation has error messages being raised (and caught) for comparing the roots and coroots. This will be done at some point on a separate PR.

### :memo: Checklist

- [x] The title is concise, informative, and self-explanatory.
- [x] The description explains in detail what this PR is about.
- [x] I have linked a relevant issue or discussion.
- [x] I have created tests covering the changes.
- [x] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by fchapoton** [comment]
some problems with imports, apparently

**Comment by tscrim** [comment]
That should take care of all of the failures.

**Comment by mkoeppe** [comment]
Needs rebase

File: src/sage/monoids/indexed_free_monoid.py [hunk 1]
```diff
@@ -961,6 +963,11 @@ class IndexedFreeAbelianMonoid(IndexedMonoid):
         sage: F = FreeAbelianMonoid(index_set=Partitions(), prefix='A', bracket=False, scalar_mult='%')
         sage: F.gen([3,1,1]) * F.gen([2,2])
         A[2, 2]%A[3, 1, 1]
+
+    .. TODO::
+
+        Implement a subclass when the index sets is finite that utilizes
+        vectors or the polydict monomials with the index order is fixed.
```
Comment (mkoeppe) [inline src/sage/monoids/indexed_free_monoid.py]:
```suggestion
        vectors or the polydict monomials with the index order fixed.
```

File: src/sage/monoids/indexed_free_monoid.py [hunk 2]
```diff
@@ -288,7 +288,11 @@ def support(self):
             [0, 2]
         """
         supp = {key for key, exp in self._sorted_items() if exp != 0}
-        return sorted(supp)
+        try:
+            return sorted(supp, key=print_options['sorting_key'],
+                          reverse=print_options['sorting_reverse'])
+        except Exception: # Sorting the output is a plus, but if we can't, no big deal
```
Comment (mkoeppe) [inline src/sage/monoids/indexed_free_monoid.py]:
```suggestion
        except Exception:  # Sorting the output is a plus, but if we can't, no big deal
```

File: src/sage/combinat/root_system/weight_space.py [hunk 3]
```diff
@@ -532,17 +532,51 @@ def is_dominant(self):
...
+            sage: w.is_dominant_weight()
+            True
+            sage: w = Lambda[1] + 2/3*Lambda[3]
+            sage: w.is_dominant_weight()
+            False
+            sage: w = Lambda[1] - Lambda[2]
+            sage: w.is_dominant_weight()
+            False
+
+        In the extended affine weight lattice, 'delta' is orthogonal to
```
Comment (mkoeppe) [inline src/sage/combinat/root_system/weight_space.py]:
```suggestion
        In the extended affine weight lattice, ``delta`` is orthogonal to
```

File: src/sage/combinat/root_system/root_lattice_realizations.py [hunk 4]
```diff
@@ -3910,6 +3910,57 @@ def is_dominant_weight(self): # Or is_dominant_integral_weight?
...
+
+            A weight `\lambda` is *Verma dominant* if
+
+            .. MATH::
+
+                \langle \lambda + \rho, \alpha^{\vee} \rangle \notin \ZZ_{<0}
+
+            for all positive roots `\alpha`. Note that this does *not*
+            imply that `\langle \lambda + \rho, \alpha^{\vee} \rangle \geq 0`
+            for any positive root `\alpha`. This is used to determine if
```
Comment (mkoeppe) [inline src/sage/combinat/root_system/root_lattice_realizations.py:3927]: What does "This" refer to in this sentence?

File: src/sage/categories/triangular_kac_moody_algebras.py [hunk 5]
```diff
@@ -262,3 +337,99 @@ def part(self):
...
+
+        class ElementMethods:
+            def transpose(self):
+                r"""
+                Return the transpose of ``self``.
+
+                The transpose `\tau` is the map that sends the root basis
+                elements `e_{\alpha} \leftrightarrow e_{-\alpha}` and fixes
+                the Cartan subalgebra `h_{\alpha}`. It is an anti-involution
+                in the sense `[\tau(a), \tau(b)] = \tau([a, b])`.
```
Comment (mkoeppe) [inline src/sage/categories/triangular_kac_moody_algebras.py]: what's "anti" here?

File: src/sage/algebras/lie_algebras/poincare_birkhoff_witt.py [hunk 6]
```diff
@@ -630,3 +637,169 @@ def _act_on_(self, x, self_on_left):
...
+        EXAMPLES::
+
+            sage: U = lie_algebras.so(QQ, 5).pbw_basis()
+            sage: TestSuite(U).run()
+
+            sage: L = lie_algebras.sl(QQ, 2)
+            sage: PBW = L.pbw_basis()
+            sage: E, F, H = PBW.algebra_generators()
+            sage: TestSuite(PBW).run(elements=[E, F, H])
+            sage: TestSuite(PBW).run(elements=[E, F, H, E*F + H]) # long time
```
Comment (mkoeppe) [inline src/sage/algebras/lie_algebras/poincare_birkhoff_witt.py]:
```suggestion
            sage: TestSuite(PBW).run(elements=[E, F, H, E*F + H])  # long time
```

File: src/sage/algebras/lie_algebras/poincare_birkhoff_witt.py [hunk 7]
```diff
@@ -630,3 +637,169 @@ def _act_on_(self, x, self_on_left):
...
+        def transpose(self):
+            r"""
+            Return the transpose map of ``self``.
+
+            This is the tranpose map on the Lie algebra extended
+            as an anti-involution of ``self``.
+
+            EXAMPLES::
+
+                sage: g = LieAlgebra(QQ, cartan_type=['D', 4])
```
Comment (mkoeppe) [inline src/sage/algebras/lie_algebras/poincare_birkhoff_witt.py:803]: expand this example?

File: src/sage/algebras/lie_algebras/verma_module.py [hunk 8]
```diff
@@ -35,7 +35,80 @@
 from sage.rings.rational_field import QQ
 
 
-class VermaModule(CombinatorialFreeModule):
+class ModulePrinting(object):
```
Comment (mkoeppe) [inline src/sage/algebras/lie_algebras/verma_module.py]:
```suggestion
class ModulePrinting:
```
`object` not needed in Python 3

File: src/sage/algebras/lie_algebras/bgg_dual_module.py [hunk 9]
```diff
@@ -0,0 +1,1174 @@
...
+        For each image vector of `f_{\beta_1}^{b_1} \cdots f_{\beta_k}^{b_k}
+        v_{\lambda}` at the current depth `b_1 + \cdots b_k`, consider the
+        image under multiplication by every generator `f_{\alpha}` for
+        all `\alpha \leq \beta_1` in the fixed PBW ordering of the (dual)
+        Verma module.
+
+        .. TODO::
+
+            Avoid unnecessary computations by using the corresponding
+            (combinastorial) crystal.
```
Comment (mkoeppe) [inline src/sage/algebras/lie_algebras/bgg_dual_module.py]:
```suggestion
            (combinatorial) crystal.
```

File: src/sage/algebras/lie_algebras/bgg_dual_module.py [hunk 10]
```diff
@@ -0,0 +1,1174 @@
...
+        EXAMPLES::
+
+            sage: g = LieAlgebra(QQ, cartan_type=['E', 6])
+            sage: La = g.cartan_type().root_system().weight_space().fundamental_weights()
+            sage: type(g.simple_module(La[1]+La[2]))
+            <class 'sage.algebras.lie_algebras.bgg_dual_module.FiniteDimensionalSimpleModule_with_category'>
+            sage: type(g.simple_module(La[1]-La[2]))
+            <class 'sage.algebras.lie_algebras.bgg_dual_module.SimpleModule_with_category'>
+            sage: type(g.simple_module(La[1]+3/2*La[2]))
+            <class 'sage.algebras.lie_algebras.bgg_dual_module.SimpleModule_with_category'>
```
Comment (mkoeppe) [inline src/sage/algebras/lie_algebras/bgg_dual_module.py]:
```suggestion
            sage: type(g.simple_module(La[1] + La[2]))
            <class 'sage.algebras.lie_algebras.bgg_dual_module.FiniteDimensionalSimpleModule_with_category'>
            sage: type(g.simple_module(La[1] - La[2]))
            <class 'sage.algebras.lie_algebras.bgg_dual_module.SimpleModule_with_category'>
            sage: type(g.simple_module(La[1] + 3/2*La[2]))
            <class 'sage.algebras.lie_algebras.bgg_dual_module.SimpleModule_with_category'>
```

File: src/sage/algebras/lie_algebras/bgg_dual_module.py [hunk 11]
```diff
@@ -0,0 +1,1174 @@
...
+
+            sage: g = LieAlgebra(QQ, cartan_type=['G', 2])
+            sage: La = g.cartan_type().root_system().weight_lattice().fundamental_weights()
+            sage: L = g.simple_module(La[1]+La[2])
+            sage: TestSuite(L).run()
+
+            sage: g = LieAlgebra(QQ, cartan_type=['A', 2])
+            sage: La = g.cartan_type().root_system().weight_lattice().fundamental_weights()
+            sage: L = g.simple_module(La[1]-La[2])
+            sage: TestSuite(L).run()
```
Comment (mkoeppe) [inline src/sage/algebras/lie_algebras/bgg_dual_module.py]:
```suggestion
            sage: g = LieAlgebra(QQ, cartan_type=['G', 2])
            sage: La = g.cartan_type().root_system().weight_lattice().fundamental_weights()
            sage: L = g.simple_module(La[1] + La[2])
            sage: TestSuite(L).run()

            sage: g = LieAlgebra(QQ, cartan_type=['A', 2])
            sage: La = g.cartan_type().root_system().weight_lattice().fundamental_weights()
            sage: L = g.simple_module(La[1] - La[2])
            sage: TestSuite(L).run()
```

File: src/sage/algebras/lie_algebras/bgg_dual_module.py [hunk 12]
```diff
@@ -0,0 +1,1174 @@
...
+    def dual(self):
+        r"""
+        Return the dual module of ``self``, which is ``self`` since simple
+        modules are self-dual.
+
+        EXAMPLES::
+
+            sage: g = LieAlgebra(QQ, cartan_type=['B', 4])
+            sage: La = g.cartan_type().root_system().weight_lattice().fundamental_weights()
+            sage: L = g.simple_module(2*La[1]+3*La[4])
```
Comment (mkoeppe) [inline src/sage/algebras/lie_algebras/bgg_dual_module.py]:
```suggestion
            sage: L = g.simple_module(2*La[1] + 3*La[4])
```

File: src/sage/algebras/lie_algebras/bgg_dual_module.py [hunk 13]
```diff
@@ -0,0 +1,1174 @@
...
+    def homogeneous_component_basis(self, mu):
+        r"""
+        Return a basis for the ``mu`` weight space of ``self``.
+
+        EXAMPLES::
+
+            sage: g = LieAlgebra(QQ, cartan_type=['A', 2])
+            sage: P = g.cartan_type().root_system().weight_lattice()
+            sage: La = P.fundamental_weights()
+            sage: la = La[1]+La[2]
```
Comment (mkoeppe) [inline src/sage/algebras/lie_algebras/bgg_dual_module.py]:
```suggestion
            sage: la = La[1] + La[2]
```

File: src/sage/algebras/lie_algebras/bgg_dual_module.py [hunk 14]
```diff
@@ -0,0 +1,1174 @@
...
+    class Element(CombinatorialFreeModule.Element):
+        def _acted_upon_(self, scalar, self_on_left=True):
+            r"""
+            Return the action of ``scalar`` on ``self``.
+
+            EXAMPLES::
+
+                sage: g = LieAlgebra(QQ, cartan_type=['A', 2])
+                sage: La = g.cartan_type().root_system().weight_lattice().fundamental_weights()
+                sage: L = g.simple_module(La[1]+La[2])
```
Comment (mkoeppe) [inline src/sage/algebras/lie_algebras/bgg_dual_module.py]:
```suggestion
                sage: L = g.simple_module(La[1] + La[2])
```

File: src/sage/algebras/lie_algebras/bgg_dual_module.py [hunk 15]
```diff
@@ -0,0 +1,1174 @@
...
+    """
+    def bgg_resolution(self):
+        """
+        Return the BGG resolution of ``self``.
+
+        EXAMPLES::
+
+            sage: g = LieAlgebra(QQ, cartan_type=['A', 2])
+            sage: La = g.cartan_type().root_system().weight_lattice().fundamental_weights()
+            sage: L = g.simple_module(La[1]+La[2])
```
Comment (mkoeppe) [inline src/sage/algebras/lie_algebras/bgg_dual_module.py]:
```suggestion
            sage: L = g.simple_module(La[1] + La[2])
```

File: src/sage/algebras/lie_algebras/bgg_resolution.py [hunk 16]
```diff
@@ -0,0 +1,229 @@
...
+        sage: L = g.simple_module(La[1]+4*La[2])
+        sage: res = L.bgg_resolution()
+        sage: ascii_art(res)
+                                [ 1 -1]       [1]
+                    [1 1]       [-1  1]       [1]
+         0 <-- C_0 <------ C_1 <-------- C_2 <---- C_3 <-- 0
+
+        sage: g = LieAlgebra(QQ, cartan_type=['D', 4])
+        sage: La = g.cartan_type().root_system().weight_lattice().fundamental_weights()
+        sage: L = g.simple_module(La[1]+La[2]+3*La[3])
```
Comment (mkoeppe) [inline src/sage/algebras/lie_algebras/bgg_resolution.py]:
```suggestion
        sage: L = g.simple_module(La[1] + La[2] + 3*La[3])
```

File: src/sage/algebras/lie_algebras/bgg_resolution.py [hunk 17]
```diff
@@ -0,0 +1,229 @@
...
+
+    INPUT:
+
+    - ``L`` -- a simple module
+
+    EXAMPLES::
+
+        sage: g = LieAlgebra(QQ, cartan_type=['A', 2])
+        sage: La = g.cartan_type().root_system().weight_lattice().fundamental_weights()
+        sage: L = g.simple_module(La[1]+4*La[2])
```
Comment (mkoeppe) [inline src/sage/algebras/lie_algebras/bgg_resolution.py]:
```suggestion
        sage: L = g.simple_module(La[1] + 4*La[2])
```

File: src/sage/algebras/lie_algebras/bgg_resolution.py [hunk 18]
```diff
@@ -0,0 +1,229 @@
...
+    """
+    def __init__(self, L):
+        r"""
+        Initialize ``self``.
+
+        EXAMPLES::
+
+            sage: g = LieAlgebra(QQ, cartan_type=['B', 2])
+            sage: La = g.cartan_type().root_system().weight_lattice().fundamental_weights()
+            sage: L = g.simple_module(La[1]+La[2])
```
Comment (mkoeppe) [inline src/sage/algebras/lie_algebras/bgg_resolution.py]:
```suggestion
            sage: L = g.simple_module(La[1] + La[2])
```

File: src/sage/algebras/lie_algebras/verma_module.py [hunk 19]
```diff
@@ -979,12 +1138,66 @@ def is_surjective(self):
...
+    def image(self):
+        r"""
+        Return the image of ``self``.
+
+        EXAMPLES::
+
+            sage: g = LieAlgebra(QQ, cartan_type=['B', 2])
+            sage: La = g.cartan_type().root_system().weight_lattice().fundamental_weights()
+            sage: M = g.verma_module(La[1]+2*La[2])
+            sage: Mp = g.verma_module(La[1]+3*La[2])
```
Comment (mkoeppe) [inline src/sage/algebras/lie_algebras/verma_module.py]:
```suggestion
            sage: M = g.verma_module(La[1] + 2*La[2])
            sage: Mp = g.verma_module(La[1] + 3*La[2])
```

File: src/sage/algebras/lie_algebras/verma_module.py [hunk 20]
```diff
@@ -1075,6 +1292,37 @@ def _an_element_(self):
...
+    def highest_weight_image(self):
+        r"""
+        Return the image of the highest weight vector of the domain
+        in the codomain.
+
+        EXAMPLES::
+
+            sage: g = LieAlgebra(QQ, cartan_type=['C', 3])
+            sage: La = g.cartan_type().root_system().weight_lattice().fundamental_weights()
+            sage: M = g.verma_module(La[1]+2*La[3])
```
Comment (mkoeppe) [inline src/sage/algebras/lie_algebras/verma_module.py]:
```suggestion
            sage: M = g.verma_module(La[1] + 2*La[3])
```

File: src/sage/algebras/lie_algebras/bgg_dual_module.py [hunk 21]
```diff
@@ -0,0 +1,1174 @@
...
+
+        EXAMPLES::
+
+            sage: g = LieAlgebra(QQ, cartan_type=['B', 2])
+            sage: La = g.cartan_type().root_system().weight_space().fundamental_weights()
+            sage: M = g.verma_module(2*La[1]+La[2])
+            sage: Mc = M.dual()
+            sage: TestSuite(Mc).run()
+
+            sage: M = g.verma_module(2/3*La[1]-3/5*La[2])
```
Comment (mkoeppe) [inline src/sage/algebras/lie_algebras/bgg_dual_module.py]:
```suggestion
            sage: M = g.verma_module(2*La[1] + La[2])
            sage: Mc = M.dual()
            sage: TestSuite(Mc).run()

            sage: M = g.verma_module(2/3*La[1] - 3/5*La[2])
```

File: src/sage/algebras/lie_algebras/bgg_dual_module.py [hunk 22]
```diff
@@ -0,0 +1,1174 @@
...
+    @staticmethod
+    def __classcall__(cls, simple, prefix='f', **kwds):
+        r"""
+        Normalize input to ensure a unique representation.
+
+        TESTS::
+
+            sage: g = LieAlgebra(QQ, cartan_type=['E', 6])
+            sage: La = g.cartan_type().root_system().weight_lattice().fundamental_weights()
+            sage: L = g.simple_module(La[1]+La[3])
```
Comment (mkoeppe) [inline src/sage/algebras/lie_algebras/bgg_dual_module.py]:
```suggestion
            sage: L = g.simple_module(La[1] + La[3])
```

File: src/sage/algebras/lie_algebras/poincare_birkhoff_witt.py (hunk 7 above)
Comment (tscrim) [reply src/sage/algebras/lie_algebras/poincare_birkhoff_witt.py:803]: Done; I just forgot to add it.

File: src/sage/categories/triangular_kac_moody_algebras.py (hunk 5 above)
Comment (tscrim) [reply src/sage/categories/triangular_kac_moody_algebras.py]: Typo: `[b, a]`.

File: src/sage/combinat/root_system/root_lattice_realizations.py (hunk 4 above)
Comment (tscrim) [reply src/sage/combinat/root_system/root_lattice_realizations.py:3927]: Clearly it should refer to being Verma dominant; I made this explicit.

**Comment by tscrim** [comment]
I have made all other suggested changes.

**Comment by mkoeppe** [comment]
Lint complains

**Comment by tscrim** [comment]
Fixed, although I don't know why the linter didn't catch that before.

**Review by mkoeppe — APPROVED** [review]
LGTM.

**Comment by tscrim** [comment]
Thank you!

**Pull request merged**